File: leetcode-practice-java/src/main/java/io/hank/leetcode/practices/_11_Container_With_Most_Water.java
package io.hank.leetcode.practices;

import io.hank.leetcode.annotations.ComplexityType;
import io.hank.leetcode.annotations.SpaceComplexity;
import io.hank.leetcode.annotations.TimeComplexity;

public class _11_Container_With_Most_Water extends LeetcodeProblemSolution {

    @TimeComplexity(ComplexityType.O_N)
    @SpaceComplexity(ComplexityType.O_1)
    int maxArea(int[] height) {
        int left = 0;
        int right = height.length - 1;
        int maxContainer = 0;

        while (left < right) {
            int currentContainer = Math.min(height[left], height[right]) * (right - left);
            maxContainer = Math.max(maxContainer, currentContainer);
            if (height[left] > height[right]) {
                right--;
            } else {
                left++;
            }
        }

        return maxContainer;
    }

}

File: leetcode-practice-typescript/src/two-pointers/11-container-with-most-water.ts
function maxArea(height: number[]): number {
    let left = 0;
    let right = height.length - 1;
    let res = Math.min(height[left], height[right]) * (right - left);
    while (left < right) {
        // 哪边小就移动哪边
        if (height[left] > height[right]) {
            const rh = height[right];
            while (left < --right && height[right] <= rh) ;
        } else {
            const lh = height[left];
            while (++left < right && height[left] <= lh) ;
        }
        res = Math.max(res, Math.min(height[left], height[right]) * (right - left));
    }
    return res;
}

console.log(maxArea([1, 8, 6, 2, 5, 4, 8, 3, 7]));
